quantization: remove most fp16 configs from fbgemm/qnnpack

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#78528 quantization: remove most fp16 configs from fbgemm/qnnpack**

Summary:

The fbgemm and qnnpack backends mostly support ops with quint8 activations.
Historically, the default backend config has included ops with fp16 activations
for other backends. This PR keeps the old config under a different name to keep
the functionality tested, and makes the default config match fbgemm/qnnpack ops.

Test plan:

```
python test/test_quantization.py -k TestQuantizeFx
```

Differential Revision: [D36778710](https://our.internmc.facebook.com/intern/diff/D36778710)

vkuzo committed:
quantization: remove most fp16 configs from fbgemm/qnnpack

Summary:

The fbgemm and qnnpack backends mostly support ops with quint8 activations.
Historically, the default backend config has included ops with fp16 activations
for other backends. This PR keeps the old config under a different name to keep
the functionality tested, and makes the default config match fbgemm/qnnpack ops.

Test plan:

```
python test/test_quantization.py -k TestQuantizeFx
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36778710).

vkuzo committed:
Update on "quantization: remove most fp16 configs from fbgemm/qnnpack"


Summary:

The fbgemm and qnnpack backends mostly support ops with quint8 activations.
Historically, the default backend config has included ops with fp16 activations
for other backends. This PR keeps the old config under a different name to keep
the functionality tested, and makes the default config match fbgemm/qnnpack ops.

Test plan:

```
python test/test_quantization.py -k TestQuantizeFx
```

Differential Revision: [D36778710](https://our.internmc.facebook.com/intern/diff/D36778710)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36778710).

vkuzo commented on torch/ao/quantization/backend_config/native.py:
```diff
@@ -682,14 +679,16 @@ def _get_embedding_op_configs():
 
-def get_native_backend_config_dict():
-    """ Get backend_config_dict for PyTorch Native backend (fbgemm/qnnpack). """
+def get_native_and_fp16_backend_config_dict():
```
I'm open to a new name, let's discuss offline

jerryzh168 commented:
qq: why don't we just remove fp16 configs?

vkuzo commented:
> qq: why don't we just remove fp16 configs?

I left it in because the goal of this PR is just to make the fbgemm/qnnpack config match reality, and I didn't want to leave existing behavior untested.  If there is a future PR which removes the ability of handling fp16 ops, that could work.  Since backends we care about support fp16, perhaps adding fp16 configs to other backend configuration dicts could work as well.

andrewor14 commented on torch/ao/quantization/backend_config/native.py:
```diff
@@ -700,6 +699,43 @@ def get_native_backend_config_dict():
+            *_get_bn_configs(),
+            *_get_share_qparams_op_configs(share_qparams_op_dtype_configs),
+            *_get_rnn_op_configs(),
+            *_get_embedding_op_configs(),
```
I think in this case it may be worth it in case we want to add/remove ops to this list, e.g. if we forget to modify one of these two places then the resulting behavior may be difficult to debug. Not a strong preference though. Either is fine with me.

vkuzo replied:
seems like a tradeoff of optimizing for reading existing code versus optimizing for writing new code.  I'd estimate the "easy to read" use case to be more important for this instance.

In fact, I'd argue that removing a lot of the utilities in this file and replacing it with dead simple and duplicated configs might be valuable in the future.  Currently, this file seems optimized for writes but it really difficult to read.

vkuzo replied:
I'd love to hear ideas on how to both enforce consistency for adding new things as well as keep things readable.  If we can only pick one, I think picking readability is a better choice in this case.

jerryzh168 commented:
> I didn't want to leave existing behavior untested

just to make sure I understand, we are keeping fp16 configs right now to make sure fp16 behavior is still tested? sgtm, just want to confirm the goal.

vkuzo committed:
Update on "quantization: remove most fp16 configs from fbgemm/qnnpack"


Summary:

The fbgemm and qnnpack backends mostly support ops with quint8 activations.
Historically, the default backend config has included ops with fp16 activations
for other backends. This PR keeps the old config under a different name to keep
the functionality tested, and makes the default config match fbgemm/qnnpack ops.

Test plan:

```
python test/test_quantization.py -k TestQuantizeFx
```

Differential Revision: [D36778710](https://our.internmc.facebook.com/intern/diff/D36778710)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36778710).

vkuzo committed:
Update on "quantization: remove most fp16 configs from fbgemm/qnnpack"


Summary:

The fbgemm and qnnpack backends mostly support ops with quint8 activations.
Historically, the default backend config has included ops with fp16 activations
for other backends. This PR keeps the old config under a different name to keep
the functionality tested, and makes the default config match fbgemm/qnnpack ops.

Test plan:

```
python test/test_quantization.py -k TestQuantizeFx
```

Differential Revision: [D36778710](https://our.internmc.facebook.com/intern/diff/D36778710)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36778710).
